### PR TITLE
Crew Objectives

### PR DESCRIPTION
**Pull request by ike709**
:cl: Citadel (of all places), ported by ike709
add: Crew objectives are now a thing. Now you can get greentext without being an antag. If, for some ridiculous reason, you don't want an objective, it can be disabled in your game preferences.
/:cl:

Before any jimmies are rustled, note that I plan on making this a player preference for if they want to have objectives or not. Also, I removed miscreant objectives because I'm not comfortable with adding Tator Lite until the greytide has been dealt with.

TODO:
- [x] Add a player preference for getting objectives.
- [x] Make sure any and all gross citadel stuff is removed.
- [x] Make sure the current objectives are a good fit for Oracle. Honestly I haven't read most of them yet.
- [x] Add objectives for Oracle's jobs.
- [x] Add more objectives in general. Accepting suggestions.
- [x] Test that this system works at all. I haven't booted up a local server yet.
- [x] Test the preference.
- [x] ~~Test the existing objectives.~~
- [x] ~~Test any and all newly added objectives.~~
- [x] ~~Test that this doesn't fuck up actual antag objectives in some weird way.~~
Edit: Honestly, it'd probably be easier to just testmerge and fix whatever breaks. It works fine for the objectives I've tried so far.

### CODE REVIEW

File: code/modules/crew_objectives/civilian_objectives.dm [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+/datum/objective/crew/vow/check_completion()
+	if(owner && owner.current)
+		var/list/say_log = owner.current.logging[INDIVIDUAL_SAY_LOG]
+		if(say_log.len > 0)
+			return FALSE
+	return TRUE
+
+/datum/objective/crew/nullrod
+	explanation_text = "Don't lose your holy rod."
```
Comment (ike709) [inline code/modules/crew_objectives/civilian_objectives.dm]: You can guess what this objective also checked for on citadel if the chaplain was male.

**Comment by Ziiro** [comment]
I guess my big question is: Is there any way to track "Number of successful objectives" on someone's account, so we can go full Achievement points dickwaving on this?

File: code/modules/crew_objectives/cargo_objectives.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*				CARGO OBJECTIVES				*/
+
+/datum/objective/crew/petsplosion
+	explanation_text = "Ensure there are at least (If you see this, yell on citadels discord in the development discussion channel) pets on the station by the end of the shift. Interpret this as you wish."
```
Comment (Xantholne) [inline code/modules/crew_objectives/cargo_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/cargo_objectives.dm [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			if(H.z == ZLEVEL_STATION_PRIMARY || SSshuttle.emergency.shuttle_areas[get_area(H)])
+				if(istype(H.wear_neck, /obj/item/clothing/neck/petcollar))
+					petcount--
+	if(petcount <= 0)
+		return TRUE
+	else
+		return FALSE
+
+/datum/objective/crew/points //ported from old hippie
+	explanation_text = "Make sure the station has at least (Something broke, report this to the development discussion channel of citadels discord) supply points at the end of the shift."
```
Comment (Xantholne) [inline code/modules/crew_objectives/cargo_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/cargo_objectives.dm [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	jobs = "quartermaster,cargotechnician"
+
+/datum/objective/crew/points/New()
+	. = ..()
+	target_amount = rand(25000,100000)
+	update_explanation_text()
+
+/datum/objective/crew/points/update_explanation_text()
+	. = ..()
+	explanation_text = "Make sure the station has at least [target_amount] supply points at the end of the shift."
```
Comment (Xantholne) [inline code/modules/crew_objectives/cargo_objectives.dm]: supply points needs to be changed to station credits

File: code/modules/crew_objectives/cargo_objectives.dm [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	jobs = "shaftminer"
+
+/datum/objective/crew/bubblegum/check_completion()
+	for(var/mob/living/simple_animal/hostile/megafauna/bubblegum/B in GLOB.mob_list)
+		if(!(B.stat == DEAD))
+			return FALSE
+	return TRUE
+
+/datum/objective/crew/fatstacks //ported from old hippie
+	explanation_text = "Have at least (something broke, report this to the development discussion channel of citadels discord) mining points on your ID at the end of the shift."
```
Comment (Xantholne) [inline code/modules/crew_objectives/cargo_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/engineering_objectives.dm [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	var/datum/station_state/end_state = new /datum/station_state()
+	end_state.count()
+	var/station_integrity = min(PERCENT(GLOB.start_state.score(end_state)), 100)
+	if(!SSticker.mode.station_was_nuked && station_integrity >= target_amount)
+		return TRUE
+	else
+		return FALSE
+
+/datum/objective/crew/poly
+	explanation_text = "Make sure Poly keeps his headset, and stays alive until the end of the shift."
```
Comment (Ziiro) [inline code/modules/crew_objectives/engineering_objectives.dm]: 👌

File: code/modules/crew_objectives/civilian_objectives.dm [hunk 7]
```diff
@@ -0,0 +1,223 @@
+/*				CIVILIAN OBJECTIVES			*/
+
+/datum/objective/crew/druglordbot //ported from old Hippie with adjustments
+	var/targetchem = "none"
+	var/datum/reagent/chempath
+	explanation_text = "Have at least (somethin broke here) harvested plants containing (report this on the development discussion channel of citadel's discord) when the shift ends."
```
Comment (Xantholne) [inline code/modules/crew_objectives/civilian_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/civilian_objectives.dm [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	for(var/mob/living/carbon/human/H in GLOB.mob_list)
+		if(H.stat == DEAD && H.drunkenness >= 80)
+			if(H.z == ZLEVEL_STATION_PRIMARY || SSshuttle.emergency.shuttle_areas[get_area(H)])
+				return FALSE
+	return TRUE
+
+/datum/objective/crew/clean //ported from old Hippie
+	var/list/areas = list()
+	var/hardmode = FALSE
+	explanation_text = "Ensure sure that (Yo, something broke. Yell about this in citadels devlopmeent discussion channel.) remain spotless at the end of the shift."
```
Comment (Xantholne) [inline code/modules/crew_objectives/civilian_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/civilian_objectives.dm (hunk 1 above)
Comment (Xantholne) [reply code/modules/crew_objectives/civilian_objectives.dm]: probably needs changed to null rod or holy artefact

File: code/modules/crew_objectives/civilian_objectives.dm [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+/datum/objective/crew/nullrod/check_completion()
+	if(owner && owner.current)
+		for(var/nullrodtypes in typesof(/obj/item/nullrod))
+			if(owner.current.check_contents_for(nullrodtypes))
+				return TRUE
+	return FALSE
+
+/datum/objective/crew/reporter //ported from old hippie
+	var/charcount = 100
+	explanation_text = "Publish at least (Yo something broke) articles containing at least (Report this to Citadels development channel) characters."
```
Comment (Xantholne) [inline code/modules/crew_objectives/civilian_objectives.dm]: Citadel error message

File: code/modules/crew_objectives/civilian_objectives.dm [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+					if(length(msg.returnBody()) >= charcount)
+						target_amount--
+	if(target_amount <= 0)
+		return TRUE
+	else
+		return FALSE
+
+/datum/objective/crew/pwrgame //ported from Goon with adjustments
+	var/obj/item/clothing/targettidegarb
+	explanation_text = "Get your grubby hands on a (Dear god something broke. Report this to Citadel's development dicussion channel)."
```
Comment (Xantholne) [inline code/modules/crew_objectives/civilian_objectives.dm]: Citadel error message

**Comment by Ziiro** [comment]
In general these are fucking great. But they need more variety. Once we get the initial ones in, I could probably come up with more.

Blueshit:
* Don't lose your gun
* Ensure the captain is on the escape shuttle

Brig phys:
* Have no bodies in genpop or the brig
* Have no dead security officers

File: code/modules/crew_objectives/medical_objectives.dm [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+/datum/objective/crew/noinfections
+	explanation_text = "Make sure there are no crew members with harmful diseases at the end of the shift."
+	jobs = "virologist"
+
+/datum/objective/crew/noinfections/check_completion()
+	for(var/mob/living/carbon/human/H in GLOB.mob_list)
+		if(!H.stat == DEAD)
+			if(H.z == ZLEVEL_STATION_PRIMARY || SSshuttle.emergency.shuttle_areas[get_area(H)])
+				if(H.check_virus() == 2)
```
Comment (Xantholne) [inline code/modules/crew_objectives/medical_objectives.dm]: I don't know if this will check if it's helpful instead of just having a virus.

**Comment by ike709** [comment]
@Ziiro no, it is not tracked

**Comment by ike709** [comment]
@Ziiro you forgot IAAs

File: code/modules/crew_objectives/security_objectives.dm [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+/datum/objective/crew/enjoyyourstay/check_completion()
+	explanation_text = "Enforce Space Law to the best of your ability."
+	if(owner && owner.current)
+		if(owner.current.stat != DEAD)
+			return TRUE
+	return FALSE
+
+/datum/objective/crew/justicecrew
+	explanation_text = "Ensure there are no innocent crew members in the brig when the shift ends."
```
Comment (ike709) [inline code/modules/crew_objectives/security_objectives.dm]: This will need to be changed.

Comment (Ziiro) [reply code/modules/crew_objectives/security_objectives.dm]: Actually, does it count the shuttle tiles that give redtext as "in the brig" as well? Could still work.

And if you really wanted to go ham: give security greentext that requires all antags redtext.

**Comment by ike709** [comment]
Security objectives work. Also, there's only one security objective.

File: code/modules/crew_objectives/security_objectives.dm (hunk 12 above)
Comment (ike709) [reply code/modules/crew_objectives/security_objectives.dm]: Changed it.

File: code/modules/crew_objectives/medical_objectives.dm (hunk 11 above)
Comment (ike709) [reply code/modules/crew_objectives/medical_objectives.dm]: It will check that it is harmful.

**Comment by ike709** [comment]
I'm going to tenta

**Comment by ike709** [comment]
I'm going to tentatively say that this is ready for review, testing, and a test merge assuming everything else goes well. The system seems to work, and every job should have at least one objective.

I highly recommend doing a testmerge before actually merging.

File: code/modules/crew_objectives/security_objectives.dm [hunk 13]
```diff
@@ -0,0 +1,31 @@
+/*				SECURITY OBJECTIVES				*/
+
+/datum/objective/crew/enjoyyourstay
+	explanation_text = "Welcome aboard. Enjoy your stay."
+	jobs = "headofsecurity,securityofficer,warden,detective"
+	var/list/edglines = list("Welcome aboard. Enjoy your stay.", "You signed up for this.", "Abandon hope.", "The tide's gonna stop eventually.", "Hey, someone's gotta do it.", "No, you can't resign.", "Security is a mission, not an intermission.")
```
Comment (deathride58) [inline code/modules/crew_objectives/security_objectives.dm]: Listen

This was removed over at Cit for good reason, the HoS committing suicide over the "abandon hope" edgline can be pretty amusing at times, but after a while ya gotta think if it's truly a good idea.

Comment (ike709) [reply code/modules/crew_objectives/security_objectives.dm]: It was indeed a bad idea, which is why I changed it from
`explanation_text = pick(edglines)`
to
`explanation_text = "Enforce Space Law to the best of your ability, and survive. [pick(edglines)]"`

File: code/modules/crew_objectives/cargo_objectives.dm [hunk 14]
```diff
@@ -0,0 +1,81 @@
+/*				CARGO OBJECTIVES				*/
+
+/datum/objective/crew/petsplosion
+	explanation_text = "Ensure there are at least (If you see this, yell on GitHub) pets on the station by the end of the shift. Interpret this as you wish."
```
Comment (FlattestGuitar) [inline code/modules/crew_objectives/cargo_objectives.dm]: So a lot of these objectives have those `yell on github` lines, but seemingly they're normal, properly coded objectives? Why are those here?

Comment (ike709) [reply code/modules/crew_objectives/cargo_objectives.dm]: Indeed.

Comment (ike709) [reply code/modules/crew_objectives/cargo_objectives.dm]: The reason I don't have a default generic error message for all of them is so we can easily tell which objective it is that's breaking.
